src/core/lib/promise/mpsc.cc

```diff
@@ -0,0 +1,419 @@
+          next, reinterpret_cast<uintptr_t>(waker) | Node::kWakerPtr,
+          std::memory_order_acq_rel)) {
+    delete waker;
+    goto retry_all;
+  }
+  return Pending{};  // pending
+}
+
+void Mpsc::PushStub() {
+  stub_.next_.store(0, std::memory_order_relaxed);
+  Node* prev = head_.exchange(&stub_, std::memory_order_acq_rel);
+  DCHECK_NE(prev, nullptr);
+  prev->next_.store(reinterpret_cast<uintptr_t>(&stub_),
+                    std::memory_order_release);
+}
+
+Mpsc::Node* Mpsc::DequeueImmediate() {
+  GRPC_LATENT_SEE_INNER_SCOPE("Mpsc::DequeueImmediate");
+  Node* tail = tail_;
+  uintptr_t next = tail->next_.load(std::memory_order_acquire);
+  if (tail == &stub_) {
+    if (next == 0) {
+      // List is (ephemerally) empty.
+      return nullptr;  // pending
+    }
+    if (next & Node::kWakerPtr) {
+      // null next waker => list closed
+      DCHECK_NE(next, Node::kWakerPtr);
+      // List is (ephemerally) empty - but we've already asked to be notified
+      // when non-empty.
+      return nullptr;  // pending
+    }
+    tail = reinterpret_cast<Node*>(next);
+    tail_ = tail;
+    next = tail->next_.load(std::memory_order_acquire);
+  }
+  if (next != 0 && (next & Node::kWakerPtr) == 0) {
+    tail_ = reinterpret_cast<Node*>(next);
+    return tail;
+  }
+  Node* head = head_.load(std::memory_order_acquire);
+  // Only the reader can close, so we should not see that here.
+  DCHECK_NE(head, nullptr);
+  if (tail != head) {
+    return nullptr;  // pending
+  }
+  // add stub to queue
+  PushStub();
+  next = tail->next_.load(std::memory_order_acquire);
+  if (next & Node::kWakerPtr) {
+    // Node still being added, and we've already asked to be notified.
+    return nullptr;  // pending
+  }
+  if (next == 0) {
+    return nullptr;  // pending
+  }
+  tail_ = reinterpret_cast<Node*>(next);
+  return tail;
+}
+
+void Mpsc::Close(bool wake_reader) {
+  auto* head = head_.exchange(nullptr, std::memory_order_relaxed);
+  if (head == nullptr) return;
+  uintptr_t next = head->next_.exchange(0, std::memory_order_acq_rel);
+  if (next & Node::kWakerPtr) {
+    Waker* waker = reinterpret_cast<Waker*>(next & ~Node::kWakerPtr);
+    DCHECK_NE(waker, nullptr);
+    if (wake_reader) waker->Wakeup();
+    delete waker;
+  } else {
+    DCHECK_EQ(next, 0);
+  }
+  Node* accepted_head = accepted_head_;
+  while (accepted_head != nullptr) {
+    auto* next = accepted_head->spsc_next_;
+    ReleaseActiveTokens(accepted_head->tokens_);
+    ReleaseTokensAndClose(wake_reader, accepted_head);
+    accepted_head = next;
+  }
+  accepted_head_ = nullptr;
+  DrainMpscFrom(wake_reader, tail_);
+  tail_ = nullptr;
+}
+
+void Mpsc::ReleaseTokens(Node* node) {
+  auto prev_queued =
+      queued_tokens_.fetch_sub(node->tokens_, std::memory_order_relaxed);
+  DCHECK_GE(prev_queued, node->tokens_);
+  ReleaseActiveTokens(node->tokens_);
+  node->Unref();
+}
+
+void Mpsc::ReleaseTokensAndClose(bool wakeup, Node* node) {
```

Comment:
Nit: the wakeup argument here appears unused

Reply:
yeah, that was a bug... fixed